sdks/python/apache_beam/runners/common.py

```diff
@@ -498,6 +550,17 @@ def _invoke_per_window(
     else:
       args_for_process, kwargs_for_process = (
           self.args_for_process, self.kwargs_for_process)
+
+    # Extract key in the case of a stateful DoFn.
+    if self.user_state_context:
+      try:
+        key, unused_value = windowed_value.value
+      except (TypeError, ValueError):
+        raise ValueError(
+            ('Input value to a stateful DoFn must be a KV tuple; instead, '
+             'got %s.') % (windowed_value.value,))
+      window, = windowed_value.windows
```

Comment:
Do we know that sliding windows were already exploded? Perhaps we should be setting self.has_windowed_inputs and relying on the window assignment from above.

Reply:
Ah oops I missed what you mean.  I set the attribute per your suggestion.